### PR TITLE
Homebrew: add threat modeling

### PR DESCRIPTION
**Pull request by woodruffw**
This adds models for 4 attacker scenarios, including their postures, mitigations, and outcomes. 

CC @di and @haydentherapper for review 🙂

### CODE REVIEW

File: proposals/build-provenance-and-code-signing-for-homebrew.md [hunk 1]
```diff
@@ -110,3 +110,55 @@ For formulae with verifiable source artifacts, the signed bottle attestation des
...
+
+These models are not intended to be exhaustive.
+
+#### Compromise of bottle storage
+
+**Scenario**: Mallory is able to circumvent Homebrew's ordinary bottle building and publishing workflows and is able to upload malicious bottles to a tap's bottle storage.
+
+**Mitigation**: With *just* upload access to the bottle storage, Mallory is unable to forge valid signatures for her malicious bottles. `brew install $compromised-bottle` fails to install due to a missing or mismatched signature, and users who attempt to install `$compromised-bottle` are loudly notified.
+
+**Outcome**: Mallory is **unable** to force an install of a compromised bottle. Mallory remains undetected.
```
Comment (haydentherapper) [inline proposals/build-provenance-and-code-signing-for-homebrew.md:134]: What do you mean by undetected? I read this as an undetected attack would be in the case of a successful compromise.

Comment (haydentherapper) [reply proposals/build-provenance-and-code-signing-for-homebrew.md:134]: Ah, as I keep reading, I see why this is mentioned. It might be worth mentioning transparency as an overall goal early in the threat model.

File: proposals/build-provenance-and-code-signing-for-homebrew.md [hunk 2]
```diff
@@ -110,3 +110,55 @@ For formulae with verifiable source artifacts, the signed bottle attestation des
...
+
+**Mitigation**: With *just* access to the bottle-building workflows, Mallory still lacks access to the signing workflow. As such, she is unable to force valid signatures for her malicious bottles and remains unable to compel malicious bottle installation as described above.
+
+**Outcome**: Mallory is **unable** to force an install of a compromised bottle. Mallory remains undetected.
+
+#### Compromise of bottle-signing workflows
+
+**Scenario**: Mallory is able to interpose or otherwise control Homebrew's authentic bottle-signing workflows. She is able to sign for malicious bottles that she controls, resulting in signatures that match a given tap's signing identity but are inauthentic.
+
+**Mitigation**: With *just* access to the bottle-signing workflows, Mallory is unable to upload her inauthentically signed bottles to the bottle storage. Her bottles *would* pass signature verification but remain inaccessible to users. By virtue of Sigstore's transparency services, Mallory loses her stealth in mounting this attack.
```
Comment (haydentherapper) [inline proposals/build-provenance-and-code-signing-for-homebrew.md:148]: Should we mention in this and the next compromise that as part of the mitigation/detection, Homebrew MUST monitor for instances of core tap identities in the log and alert the ecosystem if unexpected examples are found?

File: proposals/build-provenance-and-code-signing-for-homebrew.md [hunk 3]
```diff
@@ -110,3 +110,55 @@ For formulae with verifiable source artifacts, the signed bottle attestation des
 ### macOS executable code signing
 
 Because this proposal only concerns bottles and not their individual executable members, it should have no effect on Homebrew's current use or any future uses of "ad-hoc" executable signatures (which are produced not for authenticity, but because macOS on Apple Silicon requires all executables to be signed).
+
+## Threat model
+
+### Scope
+
+The scope of this proposal is *authenticated build provenance* for Homebrew bottles.
```
Comment (haydentherapper) [inline proposals/build-provenance-and-code-signing-for-homebrew.md:118]: Should we mention somewhere that source compromise is out of scope?

**Review by di — APPROVED** [review]

File: proposals/build-provenance-and-code-signing-for-homebrew.md (hunk 1 above)
Comment (woodruffw) [reply proposals/build-provenance-and-code-signing-for-homebrew.md:134]: Yep, I meant in a transparency sense. Thanks for pointing that out, I can emphasize that!

File: proposals/build-provenance-and-code-signing-for-homebrew.md (hunk 3 above)
Comment (woodruffw) [reply proposals/build-provenance-and-code-signing-for-homebrew.md:118]: Yeah, I'll include that in this section.

File: proposals/build-provenance-and-code-signing-for-homebrew.md (hunk 2 above)
Comment (woodruffw) [reply proposals/build-provenance-and-code-signing-for-homebrew.md:150]: Done: I've put this in the summary with references to the relevant attacker models. Let me know if you think the language isn't strong enough; I couched it slightly since monitoring work isn't included in the current scope (although I don't think it'd be difficult to add) 🙂

File: proposals/build-provenance-and-code-signing-for-homebrew.md (hunk 3 above)
Comment (woodruffw) [reply proposals/build-provenance-and-code-signing-for-homebrew.md:118]: Done.

File: proposals/build-provenance-and-code-signing-for-homebrew.md (hunk 1 above)
Comment (woodruffw) [reply proposals/build-provenance-and-code-signing-for-homebrew.md:136]: Added another summary point for this, emphasizing that *just* being undetected isn't a "win" state for Mallory (she must also successfully perform her attack). Let me know if you think I should elaborate further!

**Review by haydentherapper — APPROVED** [review]
This looks great!

**Comment by di** [comment]
@woodruffw OK to merge here?

**Comment by woodruffw** [comment]
> @woodruffw OK to merge here?

OK by me, I think @trishankatdatadog may have some feedback coming. But if you want to merge first, I can address his feedback in a follow-up.

**Comment by trishankatdatadog** [comment]
> > @woodruffw OK to merge here?
> 
> OK by me, I think @trishankatdatadog may have some feedback coming. But if you want to merge first, I can address his feedback in a follow-up.

LGTM, thanks

**Comment by woodruffw** [comment]
Good to go now, @di!

**Comment by trevrosen** [comment]
Love it - thanks, @woodruffw!
